iib/workers/tasks/build_merge_index_image.py

```diff
@@ -0,0 +1,366 @@
+
+
+__all__ = ['handle_merge_request']
+
+log = logging.getLogger(__name__)
+
+
+def _create_intermediate_index_image(
+    binary_image,
+    from_index,
+    ocp_version,
+    base_dir,
+    request_id,
+    arch,
+    overwrite_from_index_token=None,
+):
+    """
+    Create index image with the same data as old one, but with new binary image and ocp version.
+
+    :param str binary_image: binary image to be used by the new index image.
+    :param str from_index: index image, whose data will be contained in the new index image.
+    :param str ocp_version: ocp version which will be added as a label to the image.
+    :param str base_dir: base directory where operation files will be located.
+    :param int request_id: the ID of the IIB build request.
+    :param str arch: the architecture to build this image for.
+    :param str overwrite_from_index_token: the token used for overwriting the input
+        ``from_index`` image. This is required for non-privileged users to use
+        ``overwrite_from_index``. The format of the token must be in the format "user:password".
+    """
+    set_request_state(request_id, 'in_progress', 'Creating intermediate index image')
+    log.info('Creating intermediate index image with a new binary image')
+    _opm_index_add(base_dir, [], binary_image, from_index, overwrite_from_index_token)
+
+    _add_ocp_label_to_index(ocp_version, base_dir, 'index.Dockerfile')
+    _build_image(base_dir, 'index.Dockerfile', request_id, arch)
+    _push_image(request_id, arch)
+    pull_spec = _create_and_push_manifest_list(request_id, [arch])
+    log.info('Intermediate image created: %s', pull_spec)
+
+
+def _add_bundles_missing_in_source(
+    source_index_bundles,
+    target_index_bundles,
+    base_dir,
+    binary_image,
+    request_id,
+    arch,
+    ocp_version,
+):
+    """
+    Rebuild index image with bundles missing from source image but present in target image.
+
+    :param list source_index_bundles: bundles present in the source index image.
+    :param list target_index_bundles: bundles present in the target index image.
+    :param str base_dir: base directory where operation files will be located.
+    :param str binary_image: binary image to be used by the new index image.
+    :param int request_id: the ID of the IIB build request.
+    :param str arch: the architecture to build this image for.
+    :param str ocp_version: ocp version which will be added as a label to the image.
+    :return: bundles which were added to the index image.
+    :rtype: list
+    """
+    set_request_state(request_id, 'in_progress', 'Adding bundles missing in source index image')
+    log.info('Adding bundles from target index image which are missing from source index image')
+    missing_bundles = []
+    source_bundle_digests = []
+    target_bundle_digests = []
+
+    for bundle in source_index_bundles:
+        if '@sha256:' in bundle['bundlePath']:
```

Comment:
IIB Already performs a similar operation when checking for already present bundles, and it doesn't display any warnings.
I suppose we could, but it might become very spammy.

If such a scenario does happen, it would be pretty big trouble, as IIB won't be guaranteed to find all the missing bundles

Reply:
@yashvardhannanavati  Thoughts?

Reply:
I added an error message if this happens.